README.md
# Terminal color

* The easiest way to style your text in the command line / terminal
* Change text color to red, green, yellow ...
* Change background color to red, green, yellow ...
* Change text style to bold, dim, underlined, blink ...
* Has support for [0 , 16, 256 and true color](asset/colors.png) ...

## Installation

Run

```
$ composer require testomat/terminal-colour
```

> Note : By default, the Windows command console doesn’t support output coloring.
> This library disables output coloring for Windows systems.
> But if your commands or scripts invoke other scripts which emit color sequences, they will be wrongly displayed as raw escape characters.
> Install the [Cmder](https://cmder.net/), [ConEmu](https://conemu.github.io/), [ANSICON](https://github.com/adoxa/ansicon/releases) or [Mintty](https://mintty.github.io/) (used by default in GitBash and Cygwin) to add coloring support to your Windows command console.

## Usage

### Using Color Styles

```php
<?php

declare(strict_types=1);

use Testomat\TerminalColour\Formatter;

$formatter = new Formatter();

// green text
$formatter->format('<info>foo</info>');

// yellow text
$formatter->format('<comment>foo</comment>');

// black text on a cyan background
$formatter->format('<question>foo</question>');

// white text on a red background
$formatter->format('<error>foo</error>');
```

The closing tag can be replaced by `</>`, which revokes all formatting options established by the last opened tag.

It’s possible to define your own styles using the Formatter class:
```php
<?php

declare(strict_types=1);

use Testomat\TerminalColour\Formatter;
use Testomat\TerminalColour\Style;

$style = new Style('red', 'yellow', ['bold', 'blink']);

$formatter = new Formatter(false, ['fire' => $style]);

$formatter->format('<fire>foo</fire>');
```

Available foreground and background colors are `black, red, green, yellow, blue, magenta, cyan, white, default, dark_grey, light_grey, light_red, light_green, light_yellow, light_blue, light_magenta, light_cyan, light_white`.

And available options are: `none, bold, dark, italic, underscore, blink, blink_fast, crossed_out, double_underline, curly_underline, overlinem, reverse` (enables the "reverse video" mode where the background and foreground colors are swapped) and `conceal` (sets the foreground color to transparent, making the typed text invisible—although it can be selected and copied; this option is commonly used when asking the user to type sensitive information).

You want to use `256 colors` or `true colors` than you must use the `StyleCode` class.

```php
<?php

declare(strict_types=1);

use Testomat\TerminalColour\Formatter;
use Testomat\TerminalColour\StyleCode;

$style256 = new StyleCode(34); // will be transformed to '38;5;34' blue
$trueStyle = new StyleCode("38;2;1;1;1"); // same goes for 1;1;1 as value; will be transformed to '38;2;1;1;1'

$formatter = new Formatter(false, ['color256Blue' => $style256,  'trueColor' => $trueStyle]);

$formatter->format('<color256Blue>foo</color256Blue>');
```

You can also set these colors and options directly inside the tag name:

```php
// green text
$formatter->format('<fg=green>foo</>');

// black text on a cyan background
$formatter->format('<fg=black;bg=cyan>foo</>');

// bold text on a yellow background
$formatter->format('<bg=yellow;options=bold>foo</>');

// bold text with underscore
$formatter->format('<options=bold,underscore>foo</>');
```

> Note: if you need to render a tag literally, escape it with a backslash: \<info> or use the escape() method to escape all the tags included in the given string.

You have the possibility with `formatAndWrap(string $message, int $width)` function to wrap you text to a specific width, 0 means no wrapping.

> Note: you want to check the color support for you terminal check the example folder out.

### Displaying Clickable Links

Commands can use the special `<href>` tag to display links like the `<a>` elements of web pages:

```php
$formatter->format('<href=https://narrowspark.com>Narrowspark Homepage</>');
```

If your terminal belongs to the [list of terminal emulators that support links][1] you can click on the "Narrowspark Homepage" text to open its URL in your default browser. Otherwise, you’ll see "Narrowspark Homepage" as regular text and the URL will be lost.

### Terminal Util
You want to know, what color or how any columns your terminal supports, than you can call:
```php
<?php

declare(strict_types=1);

use Testomat\TerminalColour\Util;

echo Util::getNumberOfColumns(); // returns the number of the columns

echo Util::getSupportedColor(); // returns the supported color as int

```

## Links
[Colors 3/4 bit](https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit)

[XVilka/TrueColour.md](https://gist.github.com/XVilka/8346728)

[Hyperlinks (a.k.a. HTML-like anchors) in terminal emulators][1]

## Versioning

This library follows semantic versioning, and additions to the code ruleset are performed in major releases.

## Changelog

Please have a look at [`CHANGELOG.md`](CHANGELOG.md).

## Contributing

Please have a look at [`CONTRIBUTING.md`](.github/CONTRIBUTING.md).

## Code of Conduct

Please have a look at [`CODE_OF_CONDUCT.md`](.github/CODE_OF_CONDUCT.md).

## License

This package is licensed using the MIT License.

Please have a look at [`LICENSE.md`](LICENSE.md).

[1]: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda


CHANGELOG.md
v1.1.1
======

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

Fixed
-----

 - [25: fix : fixed version inside composer.json](https://github.com/testomat/terminal-colour/pull/25) thanks to @prisis

v1.1.0
======

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

Added
-----

 - [23: feat : adding new utils for get number of columns and to check if ter&hellip;](https://github.com/testomat/terminal-colour/pull/23) thanks to @prisis

v1.0.0
======

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

Added
-----

 - [18: First version of testomat/terminal-colour](https://github.com/testomat/terminal-colour/pull/18) thanks to @prisis



example/terminal.php
<?php

declare(strict_types=1);

/**
 * Copyright (c) 2020 Daniel Bannert
 *
 * For the full copyright and license information, please view
 * the LICENSE.md file that was distributed with this source code.
 *
 * @see https://github.com/testomat/terminal-colour
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Testomat\TerminalColour\Formatter;
use Testomat\TerminalColour\StyleCode;
use Testomat\TerminalColour\Util;

$effects = [
    'none' => 0,
    'bold' => 1,
    'dark' => 2,
    'italic' => 3,
    'underscore' => 4,
    'blink' => 5,
    'blink_fast' => 6, // Limited support
    'reverse' => 7,
    'conceal' => 8,
    'crossed_out' => 9,
    'double_underline' => 21,
    'curly_underline' => '4:3', // Limited support
    'overline' => 53, // Limited support
];
$fgColors = [
    'black',
    'red',
    'green',
    'yellow',
    'blue',
    'magenta',
    'cyan',
    'white',
    //    'default',
    //    'dark_grey',
    'light_grey',
    'light_red',
    'light_green',
    'light_yellow',
    'light_blue',
    'light_magenta',
    'light_cyan',
    'light_white',
];

$styles = [];

for ($i = 0; $i <= 255; $i++) {
    $styles["fg_{$i}"] = new StyleCode($i);
}

for ($i = 0; $i <= 255; $i++) {
    $styles["bg_{$i}"] = new StyleCode(null, $i);
}

for ($green = 0; $green <= 255; $green += 51) {
    for ($red = 0; $red <= 255; $red += 51) {
        for ($blue = 0; $blue <= 255; $blue += 51) {
            $styles["fg_{$red};{$green};{$blue}"] = new StyleCode("38;2;{$red};{$green};{$blue}");
        }
    }
}

for ($green = 0; $green <= 255; $green += 51) {
    for ($red = 0; $red <= 255; $red += 51) {
        for ($blue = 0; $blue <= 255; $blue += 51) {
            $styles["bg_{$red};{$green};{$blue}"] = new StyleCode(null, "48;2;{$red};{$green};{$blue}");
        }
    }
}

for ($gray = 8; $gray < 256; $gray += 10) {
    $styles["fg_{$gray};{$gray};{$gray}"] = new StyleCode("38;2;{$gray};{$gray};{$gray}");
}

for ($gray = 8; $gray < 256; $gray += 10) {
    $styles["bg_{$gray};{$gray};{$gray}"] = new StyleCode(null, "48;2;{$gray};{$gray};{$gray}");
}

foreach ($effects as $n => $i) {
    $styles["effect_{$i}"] = new StyleCode(null, null, [$n]);
}

$styles['effect_21'] = new StyleCode(null, null, ['double_underline']);
$styles['effect_4:3'] = new StyleCode(null, null, ['curly_underline']);
$styles['effect_53'] = new StyleCode(null, null, ['overline']);

$color = Util::getSupportedColor();
$isColorSupported = $color !== 0;

$formatter = new Formatter($isColorSupported, $styles);

// Start Effects
echo \PHP_EOL . 'Effects:' . \PHP_EOL . \PHP_EOL;

foreach ($effects as $name => $i) {
    $text = str_pad('effect_' . $i, 14, ' ', \STR_PAD_BOTH);

    echo $formatter->format("<effect_{$i}>{$text}</>") . ' [' . str_pad((string) $i, 4, ' ', \STR_PAD_LEFT) . '] ' . $name . (in_array($i, [6, 9, 21], true) ? $formatter->format('<effect_2> (Not widely supported)</effect_2>') : '');

    echo \PHP_EOL;
}
// End Effects

// Start Colors
echo \PHP_EOL . sprintf('Colors are supported: %s' . \PHP_EOL, $isColorSupported ? 'YES' : 'NO');

$isColor256Supported = $color >= Util::COLOR256_TERMINAL;

echo \PHP_EOL . '16 colors are supported: ' . ($isColor256Supported ? 'YES' : 'NO') . \PHP_EOL;

$isColor256Supported = $color >= Util::COLOR256_TERMINAL;

echo \PHP_EOL . '256 colors are supported: ' . ($isColor256Supported ? 'YES' : 'NO') . \PHP_EOL;

if ($isColor256Supported) {
    for ($i = 0; $i <= 1; $i++) {
        $type = $i === 0 ? 'fg' : 'bg';

        echo \PHP_EOL . ' System colors:' . \PHP_EOL . \PHP_EOL;

        for ($color = 0; $color <= 8; $color++) {
            echo $formatter->format("<{$type}_{$color}>::</>");
        }

        echo \PHP_EOL;

        for ($color = 8; $color <= 16; $color++) {
            echo $formatter->format("<{$type}_{$color}>::</>");
        }

        echo \PHP_EOL;

        echo \PHP_EOL . ' Color cube, 6x6x6:' . \PHP_EOL . \PHP_EOL;

        for ($green = 0; $green <= 5; $green++) {
            for ($red = 0; $red <= 5; $red++) {
                for ($blue = 0; $blue <= 5; $blue++) {
                    $color = 16 + ($red * 36) + ($green * 6) + $blue;

                    echo $formatter->format("<{$type}_{$color}>::</>");
                }

                echo "\x1b[0m ";
            }

            echo \PHP_EOL;
        }

        echo \PHP_EOL . ' Grayscale ramp:' . \PHP_EOL . \PHP_EOL;

        for ($c = 232; $c < 256; $c++) {
            echo $formatter->format("<{$type}_{$c}>::</>");
        }

        echo \PHP_EOL;
    }
}

$isTrueColorSupported = $color >= Util::TRUECOLOR_TERMINAL;

echo \PHP_EOL . 'true colors are supported: ' . ($isTrueColorSupported ? 'YES' : 'NO') . \PHP_EOL;

if ($isTrueColorSupported) {
    echo \PHP_EOL . 'Examples for the 3-byte color mode: ' . \PHP_EOL;

    for ($i = 0; $i <= 1; $i++) {
        $type = $i === 0 ? 'fg' : 'bg';

        echo \PHP_EOL . ' Color cube:' . \PHP_EOL . \PHP_EOL;

        for ($green = 0; $green < 256; $green += 51) {
            for ($red = 0; $red < 256; $red += 51) {
                for ($blue = 0; $blue < 256; $blue += 51) {
                    echo $formatter->format("<{$type}_{$red};{$green};{$blue}>::</>");
                }

                echo "\x1b[0m ";
            }

            echo \PHP_EOL;
        }

        echo \PHP_EOL . ' Grayscale ramp:' . \PHP_EOL . \PHP_EOL;

        for ($gray = 8; $gray < 256; $gray += 10) {
            echo $formatter->format("<{$type}_{$gray};{$gray};{$gray}>::</>");
        }

        echo \PHP_EOL;
    }
}

echo \PHP_EOL;
